src/Npgsql/NpgsqlConnector.cs

```diff
@@ -1442,11 +1459,24 @@ internal void CancelRequest(bool throwExceptions = false)
                     var socketException = e.InnerException as SocketException;
                     if (socketException == null || socketException.SocketErrorCode != SocketError.ConnectionReset)
                     {
+                        cancelImmediately = true;
                         Log.Debug("Exception caught while attempting to cancel command", e, Id);
                         if (throwExceptions)
                             throw;
                     }
                 }
+                finally
+                {
+                    // If the cancellation request was not requested by a user
```

Comment:
An interesting idea.

Not sure if it would work now, with removal of the CommandCts. I'll have to think about that.

Reply:
I don't think this changes anything - if CancelRequest always triggers cancellation of the buffer's cancellation token, the next read will always fail and we end up in ReadMessageLong again, breaking the connector. We can also look at this idea after this PR.

Reply:
In fact, something similar may even be possible for the other path (where cancel doesn't fail). For user cancellation we do ReadBuffer.TimeoutCts.CancelAfter (because the operation is still ongoing in parallel), whereas for timeout cancellation we set ReadBuffer.Timeout up the stack (because the operation isn't ongoing). It's probably possible to refactor things to just make a single call on ReadBuffer that handles both these cases.

But this is really just a note for something to explore after this PR.